src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -1025,8 +1026,25 @@ async_manifest_view::size_based_retention(size_t size_limit) noexcept {
                 // The end condition is the lambda returned true, otherwise
                 // we should keep scanning.
                 auto eof = cursor->with_manifest(
```

Comment:
I'm finding it hard to reconcile the fact that we start the cursor at the archive start offset instead of the clean offset.

The main place the logic breaks down is in the fact that `cloud_log_size` uses the clean offset, but the archive start offset may be higher than the clean offset. This means since it's based on the clean offset, `to_remove` may be an overestimate, right?

As an example, let's say our offsets were

```
archive clean: 0
archive start: 500
archive_end: 1000
```

where each offset is 1 byte, for the sake of simplifying reasoning about size. Then taking an example where size-based retention is for 900 bytes:

```
size-based retention: 500
cloud_log_size: 1000
to_remove: 500 (1000 - 500)
```

and I think we'd proceed to incorrectly truncate offsets [500, 1000].

I think we should be instantiating the cursor at the clean offset, even if that means having some wasted work.

Reply:
Alternatively, maybe we should avoid computing size-based retention if `archive clean offset != archive start offset`.

Reply:
> I'm finding it hard to reconcile the fact that we start the cursor at the archive start offset instead of the clean offset.

The code is a bit misleading in this regard. The cursor is over manifests, and not segments. In terms of segments, the cursor actually starts from the first segment of the spillover manifest which contains the archive start offset.

In your example, it all works fine as long as the clean and start offsets for the archive are within the same manifest. If that's not the case, then you're right and we would delete more segments than required. Good catch. I will update retention to start from the manifest containing the clean offset and add a test for this case.

Reply:
Hmm. Thinking about this more, I think the code in `dev` can also remove more than it should.

In the example below, we have 2 spillover manifests with 4 segments each. The archive clean offset (`C`),
is at the start of the third segment in the first spill and the archive start offset (`S`) is at the start of the second segment in the second spill manifest.
```
     C         S       
|_|_|_|_| |_|_|_|_| |_|_|
^           ^            ^
Spill 1   Spill 2    STM
```

Assume all segments have the same size. Let's say retention is 4 segments (`retention = 4`). At this point the cloud log size is of 8 segments (`cloud_log_size=8`). Retention thinks 4 segments should be removed (`to_remove=cloud_log_size - retention=4`). With the code in `dev`, we start counting from the first segment in spillover 2. The truncation point we get is the start of the STM manifest. This is incorrect as we only end up keeping 2 segments.